File: MarkdownAndroid/SimpleMarkdownParserLib/src/main/java/com/crescentflare/simplemarkdownparser/symbolfinder/SimpleMarkdownSymbolStorage.java
package com.crescentflare.simplemarkdownparser.symbolfinder;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Simple markdown parser library: symbol storage
 * Stores markdown symbols being found during text scanning
 */
public class SimpleMarkdownSymbolStorage
{
    // --
    // Members
    // --

    @NotNull public List<MarkdownSymbol> symbols = new ArrayList<>();


    // --
    // Storage
    // --

    public void addSymbol(@NotNull MarkdownSymbol symbol) {
        symbols.add(symbol);
    }

    public void clearSymbols() {
        symbols = new ArrayList<>();
    }

    public void sort() {
        Collections.sort(symbols);
    }


    // --
    // Cleaning
    // --

    public void cleanOverlaps() {
        // Collect indices to remove
        List<Integer> removeSymbols = new ArrayList<>();

        // Unordered list items override text style
        for (int index = 0; index < symbols.size(); index++) {
            if (symbols.get(index).type == MarkdownSymbol.Type.FirstTextStyle && symbols.get(index).endPosition - symbols.get(index).startPosition == 1) {
                int startIndex = index + 1;
                for (int checkIndex = index - 1; checkIndex >= 0; checkIndex--) {
                    if (symbols.get(checkIndex).startPosition == symbols.get(index).startPosition) {
                        startIndex = checkIndex;
                    } else {
                        break;
                    }
                }
                for (int checkIndex = startIndex; checkIndex < symbols.size(); checkIndex++) {
                    if (checkIndex != index) {
                        if (symbols.get(checkIndex).startPosition == symbols.get(index).startPosition) {
                            if (symbols.get(checkIndex).type == MarkdownSymbol.Type.UnorderedListItem){
                                removeSymbols.add(index);
                                break;
                            }
                        } else {
                            break;
                        }
                    }
                }
            }
        }

        // Remove symbols
        for (int i = removeSymbols.size() - 1; i >= 0 ; i--) {
            int removeIndex = removeSymbols.get(i);
            symbols.remove(removeIndex);
        }
    }
}

File: MarkdownIOS/SimpleMarkdownParser/Classes/SymbolFinder/SimpleMarkdownSymbolStorage.swift
//
//  SimpleMarkdownSymbolStorage.swift
//  SimpleMarkdownParser Pod
//
//  Library symbol parsing: stores markdown symbols being found during text scanning
//

// Stores found markdown symbol markers
public class SimpleMarkdownSymbolStorage {
    
    // --
    // MARK: Members
    // --

    public var symbols = [MarkdownSymbol]()
    

    // --
    // MARK: Storage
    // --

    public func addSymbol(_ symbol: MarkdownSymbol) {
        symbols.append(symbol)
    }
    
    public func clearSymbols() {
        symbols = [MarkdownSymbol]()
    }
    
    public func sort() {
        symbols.sort { $0.startPosition < $1.startPosition }
    }
    

    // --
    // MARK: Cleaning
    // --

    public func cleanOverlaps() {
        // Collect indices to remove
        var removeIndices = [Int]()
        
        // Unordered list items override text style
        for index in symbols.indices {
            if symbols[index].type == .firstTextStyle && symbols[index].endPosition - symbols[index].startPosition == 1 {
                var startIndex = index + 1
                for checkIndex in (0..<index).reversed() {
                    if symbols[checkIndex].startPosition == symbols[index].startPosition {
                        startIndex = checkIndex
                    } else {
                        break
                    }
                }
                for checkIndex in startIndex..<symbols.count {
                    if checkIndex != index {
                        if symbols[checkIndex].startPosition == symbols[index].startPosition {
                            if symbols[checkIndex].type == .unorderedListItem {
                                removeIndices.append(index)
                                break
                            }
                        } else {
                            break
                        }
                    }
                }
            }
        }
        
        // Remove items with indices
        for index in removeIndices.reversed() {
            symbols.remove(at: index)
        }
    }
    
}
